src/Commands/core/ArchiveDumpCommands.php

```diff
@@ -0,0 +1,593 @@
+        $this->filesystem->mkdir($databaseArchiveDir);
+
+        $options['result-file'] = Path::join($databaseArchiveDir, self::SQL_DUMP_FILE_NAME);
+        $sql = SqlBase::create($options);
+        if (false === $sql->dump()) {
+            throw new Exception(dt('Unable to dump database. Rerun with --debug to see any error message.'));
+        }
+
+        return $databaseArchiveDir;
+    }
+
+    /**
+     * Returns the list of regular expressions to match paths.
+     *
+     * @param array $paths
+     *   The list of paths to match.
+     *
+     * @return array
+     */
+    private function getRegexpsForPaths(array $paths): array
+    {
+        return array_map(
+            fn($path) => sprintf('#^%s$#', addslashes(trim($path))),
+            $paths
+        );
+    }
+
+    /**
+     * Returns docroot directory name with trailing escaped slash for a "web" docroot site for use in regular expressions, otherwise - empty string.
+     *
+     * @return string
+     *
+     * @throws \Exception
+     */
+    private function getDocrootRegexpPrefix(): string
+    {
+        return $this->isWebRootSite() ? basename($this->getRoot()) . '/' : '';
+    }
+
+    /**
+     * Returns the list of regular expressions to match Drupal files paths and sites/@/settings.@.php files.
+     *
+     * @return array
+     *
+     * @throws \Exception
+     */
+    private function getDrupalExcludes(): array
+    {
+        $excludes = [
+            '#^' . $this->getDocrootRegexpPrefix() . 'sites/.+/settings\..+\.php$#',
+        ];
+
+        $drupalFilesPath = $this->getDrupalFilesDir();
+        $drupalFilesPathRelative = Path::makeRelative($drupalFilesPath, $this->getComposerRoot());
+        $excludes[] = '#^' . $drupalFilesPathRelative . '$#';
+
+        return $excludes;
+    }
+
+    /**
+     * Validates files for sensitive data (database connection).
+     *
+     * Prevents creating a code archive containing a [docroot]/sites/@/settings.php file with database connection settings
+     * defined.
+     *
+     * @param string $file
+     *   The absolute path to the file.
+     * @param string $localFileName
+     *   The local (project-base) path to the file.
+     *
+     * @throws \Exception
+     */
+    private function validateSensitiveData(string $file, string $localFileName): void
+    {
+        $regexp = '#^' . $this->getDocrootRegexpPrefix() . 'sites/.*/settings\.php$#';
+        if (!preg_match($regexp, $localFileName)) {
+            return;
+        }
+
+        // Lookup for non-empty $databases value in a site/*/settings.php file.
+        // Remove all PHP "include" and "require" directives from settings.php file before evaluating
+        // the $databases variable.
+        $settingPhpWithoutIncludes = preg_replace('/\n*\s*(include|require).+?;|<\?php/m', '', file_get_contents($file));
+
+        try {
+            eval($settingPhpWithoutIncludes);
```

Comment:
Maybe it would be better to just do a preg_match and try to find not-commented-out lines that contain `$databases =`. This might not be perfectly reliable, but we can't predict what `eval` might do on an arbitrary settings.php file.

Reply:
In Drush 8, archive:restore provides the option to restore to the database credentials stored in a settings.php file in an archive dump. This doesn't seem like a good idea to me (storing credentials in the archive), but I suppose there are use cases for it (e.g. backup and restore).

@weitzman: Do you have an opinion on whether or not we allow database credentials in the archive? Maybe we should just remove the seatbelts.

Reply:
`eval()` has been removed